Windows 11 Notepad: announce status bar if shown

josephsl opened the pull request:
### Link to issue number:
Fixes #13688 

### Summary of the issue:
With the redesigned Windows 11 Notepad, users cannot obtain statsu bar information.

### Description of how this pull request fixes the issue:
Introduce a new Windows 11 Notepad app module (notepad.exe). With the new UI design, status bar cannot be obtained as in prior Notepad releases. Thankfully it is possible to traverse the UIA tree to obtain the status bar object from a specific child element position (this is because child elements have the same UIA class name). Also, status bar is shown if focused on the document window (rich edit) and enabled from view menu.

### Testing strategy:
Manual testing:

1. With Windows 11 installed, install Notepad 11.2203 or later.
2. Open Notepad.
3. Select View/status bar.
4. Type text.
5. Perform status bar command (NVDA+End in desktop layout/NVDA+Shift+End in laptop layout).

### Known issues with pull request:
None, although status bar position may need to be changed if Microsoft updates Notepad interface in future releases.

### Change log entries:
Bug fixes:

NVDA will announce status bar content in Windows 11 Notepad. (#13386 

### Code Review Checklist:

<!--
This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review of this pull-request.
Check items to confirm you have thought about the relevance of the item.
Where items are missing (eg unit / system tests), please explain in the PR.
To check an item `- [ ]` becomes `- [x]`, note spacing.
You can also check the checkboxes after the PR is created.
A detailed explanation of this checklist is available here:
https://github.com/nvaccess/nvda/blob/master/devDocs/githubPullRequestTemplateExplanationAndExamples.md#code-review-checklist
-->

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English

josephsl committed:
appModules/Notepad (Windows 11): add a new Notepad module for Windows 11 that allows NVDA to announce status bar text if shown. Re #13688.

Introduce a new Windows 11 Notepad app module (notepad.exe). With the new UI design, status bar cannot be obtained as in prior Notepad releases. Thankfully it is possible to traverse the UIA tree to obtain the status bar object from a specific child element positiion (this is because child elements have the same UIA class name). Also, statsu bar is shown if focused on the document window (rich edit) and enabled from view menu.

seanbudd commented on source/appModules/notepad.py:
```diff
@@ -0,0 +1,31 @@
+		# Thankfully, of all the UIA objects encountered, document window has a unique window class name.
+		if api.getFocusObject().windowClassName != "RichEditD2DPT":
+			raise NotImplementedError()
+		# Look for a specific child as some children report the same UIA properties such as class name.
+		statusBar = api.getForegroundObject().children[7].firstChild
```
Could you add a brief note on how you determined that this is the index? Is there any MS documentation for this?

```suggestion
		# Look for a specific child as some children report the same UIA properties such as class name.
		notepadStatusBarIndex = 7
		statusBar = api.getForegroundObject().children[notepadStatusBarIndex].firstChild
```

seanbudd commented on source/appModules/notepad.py:
```diff
@@ -0,0 +1,31 @@
+
+class AppModule(appModuleHandler.AppModule):
+
+	def _get_statusBar(self):
```
Could you add a docstring here - here's what comes from super()

```suggestion
	def _get_statusBar(self):
		"""
		Retrieve the status bar object of the application.
		If C{NotImplementedError} is raised, L{api.getStatusBar} will resort to
		a lookup by position.
		"""
```

josephsl replied:
Done - I'll also add a note that status bar child position must be checked whenever Microsoft publishes Notepad updates to stable Windows 11 build users (Notepad 11 is updated through Microsoft Store as opposed to being tied to a Windows release).

seanbudd replied:
Can this aspect be included? It's helpful to understand what raising `NotImplementedError` will do in this function.

```
		If C{NotImplementedError} is raised, L{api.getStatusBar} will resort to
		a lookup by position.
```

josephsl committed:
Merge branch 'master' into i13688win11Notepad

josephsl committed:
appModules/notepad: add an explanation about status bar index. Re #13688.

Advised by Sean Budd (NV Access): add an explanation and a variable about Notepad 11 status bar (look for a specific UIA element (child of the foreground window) that hosts status bar text if visible).

josephsl committed:
appModules/notepad: add docstring for status bar method. Re #13688.

Advised by Sean Budd (NV Access): add a docstring for status bar retrieval method to highlight:
* The need for this routine
* Needing to check status bar child position whenver Notepad is updated (Windows 11 NOtepad is updated through Microsoft Store as opposed to being tied to a specific Windows release).

josephsl committed:
appModules/notepad: add a note about api.getStatusBar behavior. Re #13688.

Advised by Sean Budd (NV Access): should the app module method fail (and raise NotImplementedError), api.getStatusBar will resort to position lookup, therefore document this aspect.

josephsl commented:
Hi, done.

seanbudd committed:
update changes

josephsl commented:
Hi, will add a note about that (a combination of manual object navigation and Python Console was used to determine the index; the older version of Notepad 11 placed the status bar somewhere different in the UIA tree, so I used the same approach then). Thanks.

From: Sean Budd ***@***.***> 
Sent: Sunday, May 15, 2022 9:39 PM
To: nvaccess/nvda ***@***.***>
Cc: Joseph Lee ***@***.***>; Author ***@***.***>
Subject: Re: [nvaccess/nvda] Windows 11 Notepad: announce status bar if shown (PR #13690)

 

@seanbudd commented on this pull request.

  _____  

In source/appModules/notepad.py <https://github.com/nvaccess/nvda/pull/13690#discussion_r873316908> :

> +            # Look for a specific child as some children report the same UIA properties such as class name.
 
+              statusBar = api.getForegroundObject().children[7].firstChild
 

Could you add a brief note on how you determined that this is the index? Is there any MS documentation for this?

⬇️ Suggested change 

-              # Look for a specific child as some children report the same UIA properties such as class name.
 
-              statusBar = api.getForegroundObject().children[7].firstChild
 
+              # Look for a specific child as some children report the same UIA properties such as class name.
 
+              notepadStatusBarIndex = 7
 
+              statusBar = api.getForegroundObject().children[notepadStatusBarIndex].firstChild
 

—
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/13690#pullrequestreview-973388102> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AB4AXEDPL5FOFNPWRKBU6CTVKHGMRANCNFSM5VYK6ACQ> .
You are receiving this because you authored the thread.Message ID: ***@***.***>
